quantum/hotp/hotp.c

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2019 Edward Shin <contact@edwardsh.in>
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright notice,
+ *     this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <string.h>
+#include <stdint.h>
+
+#include "hotp.h"
+#include "hotp_config.h"
+#include "crypto/hmac_sha1.h"
+
+#include "tmk_core/common/eeprom.h"
+#include "quantum.h" // for send_string()
+
+#ifdef HOTP_ENABLE
+
+#ifndef HOTP_EEPROM_ADDR
+    #error HOTP_EEPROM_ADDR not defined
+#endif
+
+#ifndef STRINGIFY
+    #define STRINGIFY(a) #a
+#endif
+
+#ifndef STRINGIFY_EXPANDED
+    #define STRINGIFY_EXPANDED(a) STRINGIFY(a)
+#endif
+
+#define HOTP_CONFIG_SECRET_BYTES   (sizeof(s_hotp_secret))
+#define HOTP_COUNTER_EEPROM_ADDR   (HOTP_EEPROM_ADDR)
+#define HOTP_COUNTER_SIZE_BYTES    (8)
```

Comment:
This should probably be stored in `/tmk_core/common/eeconfig.h` rather than here. 

This is so to ensure that there are no collisions, like we've recently have had with VIA.

Reply:
Gotcha, I'll move these over there. Is there a standard format to indicate the length of the data at an address?

Reply:
the type of uint used usually specifies that, but in this case, you'd want to specify a couple of blocks.  You can add a comment to where the end is, if you'd like (eg, the next available block)